client/api/request.js

```diff
@@ -27,22 +29,39 @@ const _request = ( url, data, method ) => {
 		request.body = JSON.stringify( data );
 	}
 
-	return fetch( baseURL + url, request ).then( ( response ) => {
-		return parseJson( response ).then( ( json ) => {
-			if ( json.success ) {
-				return json;
-			}
+	const poll = ( func ) => new Promise( ( resolve, reject ) => {
+		setTimeout( () => func().then( resolve ).catch( reject ), ASYNC_POLLING_INTERVAL );
+	} );
+
+	return ( function performRequest() {
+		return fetch( baseURL + url, request ).then( ( response ) => {
+			return parseJson( response ).then( ( json ) => {
+				if ( json.async_token ) {
+					request.body = JSON.stringify( { ...( data || {} ), async_token: json.async_token } );
+					return poll( performRequest );
+				}
+				switch ( json.async_status ) {
+					case 'IN_PROGRESS':
+						return poll( performRequest );
+					case 'NOT_FOUND':
+						throw { code: 'async_token_not_found', message: __( 'The request expired' ) }; // TODO: How to phrase this error?
+				}
 
-			if ( 'rest_cookie_invalid_nonce' === json.code ) {
-				window.persistState = true;
-				alert( __( 'There was a problem saving your settings. Please try again after the page is reloaded.' ) );
-				location.reload();
-				return;
-			}
+				if ( json.success ) {
+					return json;
+				}
 
-			throw json;
+				if ( 'rest_cookie_invalid_nonce' === json.code ) {
+					window.persistState = true;
+					alert( __( 'There was a problem saving your settings. Please try again after the page is reloaded.' ) );
+					location.reload();
```

Comment:
Hum... edge case indeed, if there is polling going on every few seconds, the `nonce` should never expire, right?

Resuming polling after refresh would be tricky because right now all the polling logic is tied to this file, it isn't added to the Redux state or persisted anywhere.